### DISCUSSION TITLE
INAV RTH flies to the moon!

### DISCUSSION THREAD

**Discussion by akurekin** [iNavFlight/inav#8096] [Q&A] [answered]
## Current Behavior

When RTH is activated the drone shoots up at full throttle to about 20 meters, then starts its way to the home position. Some times it happens and some time not, more often at the wind. 
Same behaviour is observed in Position Hold when the throttle is set to full down and then, near the land, at full up.    

## Steps to Reproduce

1. Take drone off the ground
2. Activate Position Hold
3. Move to about 10 meters up
4. Throttle down
5. At about3 meters above the ground throttle up to max

## Expected behavior

Drone starts ascending at slow speed

## Suggested solution(s)

It looks like a bug. At the moment when the drone shoots up to the skies, navPos[2] shows the drone goes down and navVel[2] shows negative velocity. At the same time baro shows that the drone goes up and acc[Z] is around 2g. This moment is shown in the log attached at 36.789 sec. of flight time.

After a large number of experiments with prop balancing, barometer covering in foam and playing with settings I found a work around by setting inav_w_z_gps_v = 0 .  

## Additional context

---

- FC Board name and vendor: MATEK 405SE
- INAV version string: 3.0.1
[LOG00135.TXT](https://github.com/iNavFlight/inav/files/8842517/LOG00135.TXT)

**Comment by stronnag** [comment]
```
$ inav_states.rb  ~/dl/LOG00135.TXT 
INAV 3.0.1, states from 2.7.0
LOG00135.TXT: INAV 3.0.1 (753c4479) MATEKF405SE
Iteration	Time(s)	Elapsed(s)	State	FltMode	navFlag
   225984	 136.5	(   0.0)	nav_state_poshold_3d_in_progress (7)	ARM|ANGLE|NAVPOSHOLD|BLACKBOX		
   298592	 173.8	(  37.3)	nav_state_idle (1)	ARM|ANGLE|BLACKBOX		
   322464	 186.1	(  49.6)	nav_state_poshold_3d_in_progress (7)	ARM|ANGLE|NAVPOSHOLD|BLACKBOX		
   355882	 203.3	(  66.8)	end of log
Disarmed by: SWITCH
```
![LOG1135](https://user-images.githubusercontent.com/158229/172141482-999fa8bb-cdb6-4f00-971f-52d5cb6c7000.png)

That the baro initialises at 8m is not helping. I doubt this is a software issue. It would be interesting to see what happens with contemporary firmware, 3.0 is no longer maintained.
 
Events like this are typically caused by external stimuli (poorly protected baro, wind, vibrations).

**Comment by akurekin** [comment]
The barometer shows some thermal drift over time, but it is very slow. The barometer is also protected from wind, The vibrations are not that bad as you can see from acc measurements. I have also tested INAV 4 and 5 and the problem is even worse with these versions.  Any ideas of what the problem can be?

**Comment by stronnag** [comment]
There was a long discussion on RC Groups some months ago, the general conclusion was high frequency vibrations, most apparent on small frames, small props, high KV motors and maybe fast loop times. 

There was a theory that there might be some clipping in the IMU.

RCG discussion starts around https://www.rcgroups.com/forums/showpost.php?p=46677523&postcount=27554 and continues in the same vein for a while.

**Comment by akurekin** [comment]
Thank you very much for the reference, I would like to know what others experience is. 
On my drone I am using relatively slow Mavic motors, so no high frequency noise as you can see from the spectrum plots. Also I cannot see any clipping in the log data. 
What really looks strange is the velosity navVel[2] is estimated as negative when the copter goes up ! The velocity estimate looks wrong.

**Comment by stronnag** [comment]
> What really looks strange is the velosity navVel[2] is estimated as negative when the copter goes up ! The velocity estimate looks wrong.

That is from the accelerometer, which is the crux of the matter.

**Comment by akurekin** [comment]
Accelerometer shows the drone goes up !

**Comment by Jetrell** [comment]
Comparing your `gyro_raw `data verses the filtered data. The `gyro's (yaw) axis `still has vibrations around 50 - 60hz.

The IMU data augments the barometric data, when abrupt changes in altitude occur.... This will cause the effect you are seeing if the sensor is blinded at certain frequencies causing corrupt data, and has nothing to do with the barometer.
Filtering is very important. But its more important to find where those vibrations are coming from.
Being that they are low in the frequency range. That would likely indicate frame resonance. But checking the motors for balance, is important too. 
Makes sure the FC stack bolts are not over tightened. This can transfer vibrations to the IMU, by over-compressing the silicone dampers.

**Reply by akurekin** [reply]
Thank you for your recommendations, I'll check if the frame resonance can be an issue. I  tried to put controller on a very soft foam with no success. It is also not clear how yaw vibrations can affect the height estimate, as they have different directions.  
But if we admit that it's a resonance issue, the resonance  may happen when the drone starts moving up when the motors get high power at the first moment to start pushing the copter up. This abrupt jump in power may cause additional vibration. It must be a way to make this transition smoother to reduce a chance of resonance at this moment  This may definitely help if vibrations are caused by propeller flutter rather than frame resonance. Not all drones are perfect and this can be taken into consideration by providing more smooth control over throttle whenever possible to avoid flutter effect.

**Reply by akurekin** [reply]
My next experiment with reinforcement of drone frame improved the situation. The drone is controlled confidently in RTL mode and does not shoot up into the sky any more. So, it's a frame resonance to blame for. It's the second time in my RC  practice I am facing the problem of resonance of carbon fibre components. It looks like carbon fibre is much more affected by vibrations than aluminium and wood. I am quite surprised of the result as the frame is 5mm thick an sturdy and is not supposed to produce resonance.

**Reply by Jetrell** [reply]
Sometimes the frame resonance can be specific to the height of the motors.. That being, the height the props sit above the frame arms. Which can cause a pulsation effect on the arms, which then travels through the frame, and set up a resonance node on or near one of the stack mounting bolts.. This is often bad luck if you experience it.

**Reply by akurekin** [reply]
This can be the case, I am using the frame for mavic motors, it most likely not tested well by the community.

**Comment by akurekin** [comment]
Thank you for your recommendations, I'll check if the frame resonance can be an issue. I  tried to put controller on a very soft foam with no success. It is also not clear how yaw vibrations can affect the height estimate, as they have different directions.  
But if we admit that it's a resonance issue, the resonance  may happen when the drone starts moving up when the motors get high power at the first moment to start pushing the copter up. This abrupt jump in power may cause additional vibration. It must be a way to make this transition smoother to reduce a chance of resonance at this moment  This may definitely help if vibrations are caused by propeller flutter rather than frame resonance. Not all drones are perfect and this can be taken into consideration by providing more smooth control over throttle whenever possible to avoid flutter effect.

**Comment by mrechte** [comment]
I don't know if it is related to this problem, but when RTH starts, Z acceleration is not constant. I tried to play with nav_mc_vel_z_p (from 100 to 70) without much success:

![Capture d’écran du 2022-06-11 11-59-50](https://user-images.githubusercontent.com/5902317/173183326-fc0ec2c5-3aed-47f1-9aa2-cb78c22f5f29.png)

**Comment by Torchi** [comment]
This recent flight  of mine seems being related ?!   When RTH is activated the drone shoots up at full throttle for tens of  meters, then starts its way to the home position. 

This quad has flown few times and particularly did ok few RTH.  Because this quad is relatively big (9" props) I am fiddling with PIDs.  Last flight, the second one of the day (btw RTH worked ok on the first flight) the D part was too way high... The sudden climbing  at full power, when I issued RTH, made the voltage go down but fortunately, it just reached about 200 m, then started coming down.  Landing is slow and quad was high, so I landed it manually.   Attached log and parameters.
[INAV_5.1.0_cli_SUPER_9_20230220_180202.txt](https://github.com/iNavFlight/inav/files/10803653/INAV_5.1.0_cli_SUPER_9_20230220_180202.txt)
[blackbox_log_2023-02-20_154236.TXT](https://github.com/iNavFlight/inav/files/10803654/blackbox_log_2023-02-20_154236.TXT)

**Comment by Jetrell** [comment]
You didn't log un-filter (raw) gyro data. But from what you have logged... It shows the overall gyro noise is relatively low in a hover. 
But as soon as the throttle increases, so does the noise. 
You did mention having set the D-term too high.. That will certainly cause oscillations, that will lead to this behavior. 

I assume you read through the rest of the information in this thread.. And now have an idea of the hardware related issues that cause this to happen?

**Reply by Torchi** [reply]
I did another experiment changing PIDs and I somewhat achieved what I was thinking of... See attached step response figure.  (Note:  I made the pitch P bigger that roll P because quad has more inertia on pitch. )

This said, with this last PIDs quad behaviour is worse...  
In the previous test the sudden rise was after RTH. 
In this last test, not only the sudden rise is after RTH but also after Hold Position !
(btw, the noise of the motors does not sound "regular").

I attach blackbox and also the table of test flight vs PIDs (last test is numbered 8, previous 7).
The frame and everything else seem stiff... 
If I do not understand how to dampen vibration I can always revert to the default PIDs (HD image was more shaky but at least quad was basically "working").
[PID - Test Flights.pdf](https://github.com/iNavFlight/inav/files/10816794/PID.-.Test.Flights.pdf)

[blackbox_log_2023-02-23_165428.TXT](https://github.com/iNavFlight/inav/files/10816795/blackbox_log_2023-02-23_165428.TXT)
![Flight 8 step response](https://user-images.githubusercontent.com/18173871/220991839-8ca74303-347e-468b-8e17-11980f3ed967.jpg)

**Reply by Jetrell** [reply]
@Torchi Have you read through some of the earlier comments I wrote in this thread last June. Concerning frame resonance ?

**Reply by Torchi** [reply]
I have read it but I am currently thinking on what to do next.  
I made already sure that FCB is not too tight, that everything else is tight, that the baro is fine, etc, 
Resonance of the carbon fibre structure may definitely play a role, but I am considering sw ways to avoid it.
When I issue RTH / Hold, the quad **determines** to climb at about full motors while reaching home position.  It follows the chaos of already noisy accs / gyros.  (... and also my personal panic for the fall of the voltage ...)  Once reached home, the landing is normal.
The few times I used the default PIDs, the RTH/Hold problem seemed not to happen, so I shall surely revert to those values to do more tests and study.

**Reply by Torchi** [reply]
btw  if I ever put a reasonable limit on climb rate for RTH/Hold cases, the problem would be made less dramatic.  For automatic navigation movements, the limit is set now to  5 m/s (nav_auto_climb_rate).  But in this cases, climb speed is more ?!...

**Reply by Jetrell** [reply]
>For automatic navigation movements, the limit is set now to 5 m/s (nav_auto_climb_rate). But in this cases, climb speed is more ?!...

Because the sensor fusion altitude data is being corrupted by IMU vibrations.. This is what causes it to accelerate vertical, faster than the set-point... Until the GPS altitude and baro altitude lineup. Then it realizes the IMU acceleration data its receiving is incorrect, then it will slow and stop.

**Reply by breadoven** [reply]
Just wondering why you have `nav_mc_hover_thr `set at 1500 (default) when the BB log shows a static hover throttle of around 1300. `nav_mc_hover_thr `should be set to 1300 otherwise the throttle will surge when you select auto hover modes.

And there seems to be pretty bad vibration at throttle over 1500 during the RTH which probably resulted in a -ve climb rate even though it was in fact doing a moon shot. Although strangely it didn't seem to do this when PosHold was used later.
![Screenshot (137)](https://user-images.githubusercontent.com/56191411/221319612-f65f1b8a-cfe9-411b-b29f-8328f2ec47c4.png)

**Reply by Torchi** [reply]
I am going to try how this idea work, next time I fly !!

**Reply by Torchi** [reply]
I have given a  try to nav_mc_hover_thr = 1300 and, in the same flight, PIDs were set more similar to the default (comparison attached). 
The moonshot problem when issuing RTH changed a bit, but was not at all solved.  
Attached the BB and  a detail screenshot of the moment of RTH start. 
(As pointed out earlier, there are high vibrations when throttle is above 1500).   Throttle rises from initial 1300 and goes soon well beyond 1500  (and current becomes more than 100 Amp, battery voltage drops etc.) 
Everything manages to go back to normal a few instants later and, say, more than 100 meters higher.  
I am now trying to isolate in detail the dynamic of the return to normal descent.

[blackbox_log_2023-03-03_152552.TXT](https://github.com/iNavFlight/inav/files/10887697/blackbox_log_2023-03-03_152552.TXT)

![detail](https://user-images.githubusercontent.com/18173871/222881335-6007a0fe-96f7-48d7-b7f0-7871af83b0b9.png)

![pid](https://user-images.githubusercontent.com/18173871/222881336-e9adce81-60f1-4e7b-a2fb-efd529acc6b6.png)

**Reply by Torchi** [reply]
I am still stuck on this problem... 
I have changed all props from resin to  FC ones.  No change.
I have improved the mounting (including silicone absorbers) of the FC and now that seems to me very good. No change. 

(I have  lowered nav_mc_hover_thr but in any case throttle start ramping up form whatever level throttle is , at the moment I command HOLD or RTH.    Just in order to have less panic I also lowered max_throttle.)

In the blackbox log I attach, there are two recordings.  

The first is a "normal flight" of yesterday where I issued HOLD and it skyrocketed as usual.  Note that during WHOLE  flight  vibrations are visible in the recorded video and the sound of motors is never truly "clean".

The second part recorded, today, is with quad lying on the table.  I removed the props: sound seems normal and frame never seems to vibrate at all.  
I do not understand  this second log.  acc[z] does not seem very noisy but other quantities are indeed noisy, as navAcc[3]...

I am excluding problems with the solidity of the frame.  Resonance is possible but.. at any throttle level ?! (just stronger with higher throttle).
I am suspecting the FCB itself now (it is a SkyStars F722 Pro3 with, I believe, mounted a BMI270 gyro) 
Only, it takes time to replace it with another unit...

[blackbox_log_2023-04-17_113724.TXT](https://github.com/iNavFlight/inav/files/11250527/blackbox_log_2023-04-17_113724.TXT)

**Reply by akurekin** [reply, marked as answer]
I have got three drones with similar problem. In the first one I solved the problem by adding rubber washers to the frame arm mounts. In the second one I had to replace the rubbish bearings in all motors. In the third one, after crash, I replaced the motors themselves. And the problem has gone. If the frequency of vibrations is too high, you will not see them in the data logs.
